### DISCUSSION TITLE
Liquidator himself eligible for Liquidation

### DISCUSSION THREAD

**Discussion by mgroovyank** [Cyfrin/foundry-full-course-cu#2314] [Q&A]
Is it possible that the liquidator is also part of DSC Protocol and he himself becomes eligible for liquidation. In that case, should he be allowed to take part as liquidator?

In the below code, we revert the transaction in the last line if health factor for liquidator is broken? Can we not check the liquidator's health factor at the start of this function itself and stop the further execution of function and save gas?

```
    function liquidate(address collateral, address user, uint256 debtToCover)
        external
        moreThanZero(debtToCover)
        nonReentrant
    {
        //checks
        uint256 startingUserHealthFactor = _healthFactor(user);
        if (startingUserHealthFactor >= MIN_HEALTH_FACTOR) {
            revert DSCEngine__HealthFactorOk();
        }

        uint256 tokenAmountFromDebtToCover = getTokenAmountFromUSD(collateral, debtToCover);
        // Give liquidator 10% bonus
        // so we are giving liquidator $110 of WETH for 100 DSC
        // we should implement a feature to liquidate protocol in case of insolvency
        // and sweep extra amounts in treasury
        uint256 bonusCollateral = (tokenAmountFromDebtToCover * LIQUIDATION_BONUS) / LIQUIDATION_PRECISION;

        uint256 totalCollateralToRedeem = tokenAmountFromDebtToCover + bonusCollateral;

        //effects
        // burn dsc
        // redeem collateral to liquidator
        //interactions
        _redeemCollateral(collateral, totalCollateralToRedeem, user, msg.sender);

        _burnDSC(user, msg.sender, debtToCover);
        uint256 endingUserHealthFactor = _healthFactor(user);
        if (endingUserHealthFactor <= startingUserHealthFactor) {
            revert DSCEngine__HealthFactorNotImproved();
        }
        _revertIfHealthFactorIsBroken(msg.sender);
    }
```

**Comment by EngrPips** [comment]
Hello @mgroovyank, When It comes to writing code, there are always many ways to do things, and yeah, you are correct about checking the `health factor` of the Liquidator before proceeding, but I think that will require us to have some extra function which will run some simulation to confirm if the Liquidator will still be healthy after the Liquidation process takes place. You might incorporate it in your code to learn and explore.

**Reply by mgroovyank** [reply]
Thanks @EngrPips for the reply. But one thing, I don't think the liquidator's health factor will change in any way due to liquidation of the user. In this context, the liquidator's health factor can change only due to change in ETH/USD conversion rate. So there is no need to run any simulation as such. We can just check his health factor based on current ETH/USD price before he can take part in the protocol as a liquidator. Let me know, if my understanding is not correct.

**Reply by EngrPips** [reply]
I need to understand what you mean by `the liquidator's health factor can change only due to change in ETH/USD conversion rate`

**Reply by mgroovyank** [reply]
By that I mean, when a liquidator perform liquidation of another user. There is no change in his own total dsc minted and total collateral deposited. The only thing which can change his health factor during this process is a fall in ETH to USD price. If value of ETH falls, only in that case his own total collateral deposited value will go down, which can bring his health factor down.

**Reply by EngrPips** [reply]
Honestly, the codebase has some things that were not considered, and it is far from being a production. So, some things can be quite confusing. If we are checking to see if the price of ETH to USD has fallen after the Liquidation process to see if the Liquidator is still healthy, then how do we check that before the Liquidation process?

**Reply by mgroovyank** [reply]
Okay. Thanks. I got it. Thanks for patiently answering my questions.

**Reply by EngrPips** [reply]
You are welcome, Friend. Keep crushing it.
